update_shotgun_users.py

```diff
@@ -0,0 +1,158 @@
+# provided at the time of installation or download, or which otherwise accompanies
+# this software in either electronic or hard copy form.
+#
+
+import argparse
+import logging
+
+from shotgun_api3 import Shotgun
+
+from sg_jira import Bridge, JiraSession
+from jira import JIRAError
+
+logger = logging.getLogger("update_shotgun_users")
+
+
+def sync_jira_users_into_shotgun(sg, jira, project_key):
+    """
+    Associates JIRA users with Shotgun users.
+
+    :param sg: Connection to Shotgun
+    :param jira: Connection to JIRA
+    :param dict project_key: Project to use to match users.
+    """
+
+    # Let's make sure the sg_jira_account_id field exists and create it
+    # if missing.
+    logger.info("Ensuring HumanUser.sg_jira_account_id exists.")
+    if "sg_jira_account_id" not in sg.schema_field_read("HumanUser"):
+        logger.info("Creating HumanUser.sg_jira_account_id.")
+        sg.schema_field_create("HumanUser", "text", "Jira Account Id")
+
+    # Make sure the JIRA project exists.
+    logger.info("Locating JIRA project %s" % project_key)
+    project = jira.project(project_key)
+
+    logger.info("Retrieving all Shotgun users")
+    users = sg.find(
+        "HumanUser",
+        # User's without email or with TBD (test users) should not be considered.
+        [["email", "is_not", None], ["email", "is_not", "TBD"]],
+        ["email", "login", "sg_jira_account_id"],
+        # We sort by user id so that we always assign users in a deterministic fashion.
+        # That is important because the script can be run multiple times over time
+        # as a company hires more and more people.
+        order=[{"field_name": "id", "direction": "asc"}],
+    )
+    # Track which emails have already been mapped. This needs to happen first because
+    # the same email can be used for multiple users. So if an email has been mapped
+    # to an account id, we'll skip all users with that email when matching.
+    # This would allow an admin to run the script once, then move some JIRA account ids
+    # from  one Shotgun account to another that has the same email and run the script
+    # again without invalidating their work.
+    mapped_emails = {
+        user["email"] for user in users if user["sg_jira_account_id"] is not None
+    }
+
+    for user in users:
+
+        # Email has already been mapped to a JIRA user, so skip it.
+        if user["email"] in mapped_emails:
+            logger.info(
+                "The email '{}' from '{}' has already been associated with a JIRA account.".format(
+                    user["email"], user["login"]
+                )
+            )
+            continue
+
+        # Let's try to find a JIRA user associated with that email for the given JIRA project.
+        jira_user = jira.find_jira_assignee_for_issue(
+            user["email"], jira_project=project
+        )
+        # If no user want found, let the user know.
+        if jira_user is None:
+            continue
```

Comment:
Should there be some kind of message sent to stdout here?

Reply:
Actually, there is. I should have pointed this out in the code.

Our JIRA API wrapper's logger emits warnings when users are not found.